Content.Server/GameObjects/Components/Items/RCDComponent.cs

```
@@ -0,0 +1,169 @@
+        }
+
+        /**
+         *
+         * Method called when the user examines this object, it'll simply add the mode that it's in to the object's description
+         * @params message = The original message from examining, like ..() in BYOND's examine
+         *
+         */
+
+        void IExamine.Examine(FormattedMessage message)
+        {
+            message.AddMarkup(Loc.GetString("It's currently in "+mode+" mode, and holds "+this.ammo+" charges."));
+        }
+
+        /**
+         *
+         * Method to handle clicking on a tile to then appropriately RCD it. This can have several behaviours depending on mode.
+         * @param eventAargs = An action event telling us what tile was clicked on. We use this to exrapolate where to place the new tile / remove the old one etc.
+         *
+         */
+
+        public void AfterAttack(AfterAttackEventArgs eventArgs)
+        {
+            var attacked = eventArgs.Attacked;
+            var mapGrid = _mapManager.GetGrid(eventArgs.ClickLocation.GridID);
+            var tile = mapGrid.GetTileRef(eventArgs.ClickLocation);
+
+            var coordinates = mapGrid.GridTileToLocal(tile.GridIndices);
+            float distance = coordinates.Distance(_mapManager, Owner.Transform.GridPosition);
+
+            if (distance > InteractionSystem.InteractionRange || ammo <= 0)
+            {
+                return;
+            }
+
+            var targetTile = (ContentTileDefinition) _tileDefinitionManager[tile.Tile.TypeId];
+            ITileDefinition desiredTile = null;
+            //Try catch to assert that _outputTile is a valid key in the tiledefinition manager, failing this check, we'll abort any tile placement.
+            try
+            {
+                desiredTile = _tileDefinitionManager[_outputTile];
+            }
+            catch (Exception e)
```

I don't know what the others feel, but I'm against catching any exceptions here. If the output tile does not exist, the RCD shouldn't silently fail.

---

Yeah this definitely shouldn't be try catch'd.

---

It's to do with outputTile being set to "walls" which causes an exception. I've been unable to find a way to check if a key exists in a dict with C#

---

> I've been unable to find a way to check if a key exists in a dict with C#

[`ContainsKey`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.containskey?view=netcore-3.1) or alternatively [`TryGetValue`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.trygetvalue?view=netcore-3.1) if you want to retrieve the value too.